File: CliqueApp/Models/EventChatMessage.swift
//
//  EventChatMessage.swift
//  CliqueApp
//
//  Created by Codex on 3/7/25.
//

import Foundation
import FirebaseFirestore

struct EventChatMessage: Identifiable, Hashable {
    let id: String
    let eventId: String
    let senderId: UserID
    let senderHandle: String
    let senderName: String
    let text: String
    let createdAt: Date
    
    init(id: String = UUID().uuidString,
         eventId: String,
         senderId: UserID,
         senderHandle: String,
         senderName: String,
         text: String,
         createdAt: Date = Date()) {
        self.id = id
        self.eventId = eventId
        self.senderId = senderId
        self.senderHandle = senderHandle
        self.senderName = senderName
        self.text = text
        self.createdAt = createdAt
    }
    
    init?(data: [String: Any]) {
        guard
            let id = data["id"] as? String,
            let eventId = data["eventId"] as? String,
            let senderName = data["senderName"] as? String,
            let text = data["text"] as? String
        else {
            return nil
        }
        
        let timestamp = data["createdAt"] as? Timestamp
        let createdAt = timestamp?.dateValue() ?? Date()
        
        self.id = id
        self.eventId = eventId
        self.senderId = data["senderId"] as? String ?? data["senderUID"] as? String ?? data["senderEmail"] as? String ?? ""
        self.senderHandle = data["senderEmail"] as? String ?? ""
        self.senderName = senderName
        self.text = text
        self.createdAt = createdAt
    }
}

File: CliqueAndroid/app/src/main/java/com/clique/app/data/model/EventChatMessage.kt
package com.clique.app.data.model

import com.google.firebase.Timestamp
import com.google.firebase.firestore.DocumentSnapshot
import java.time.Instant

data class EventChatMessage(
    val id: String = "",
    val eventId: String = "",
    val senderId: String = "",
    val senderHandle: String = "",
    val senderName: String = "",
    val text: String = "",
    val createdAt: Instant = Instant.now()
) {
    companion object {
        fun fromSnapshot(snapshot: DocumentSnapshot): EventChatMessage? {
            val data = snapshot.data ?: return null
            val id = data["id"] as? String ?: snapshot.id
            val eventId = data["eventId"] as? String ?: ""
            val senderName = data["senderName"] as? String ?: ""
            val text = data["text"] as? String ?: ""
            
            val timestamp = data["createdAt"] as? Timestamp
            val createdAt = timestamp?.toDate()?.toInstant() ?: Instant.now()
            
            val senderId = data["senderId"] as? String 
                ?: data["senderUID"] as? String 
                ?: data["senderEmail"] as? String 
                ?: ""
            val senderHandle = data["senderEmail"] as? String ?: ""
            
            return EventChatMessage(
                id = id,
                eventId = eventId,
                senderId = senderId,
                senderHandle = senderHandle,
                senderName = senderName,
                text = text,
                createdAt = createdAt
            )
        }
    }
}

